### PR TITLE
Parser and serializer

### PR DESCRIPTION
**Pull request by mat-hek**

### CODE REVIEW

File: lib/membrane_opus/parser.ex [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+           PacketUtils.skip_frame_sizes(mode, data, max(0, frames - 1)),
+         new_header_size = byte_size(payload) - byte_size(rest),
+         {:ok, frames_size, data} <- PacketUtils.skip_frame_sizes(mode, data, frames),
+         body_size = frames_size + padding,
+         <<body::binary-size(body_size), data::binary>> <- data do
+      <<new_header::binary-size(new_header_size), _rest::binary>> = payload
+      parse_packets(data, [{new_header <> body, frame_duration * frames} | acc])
+    else
+      _end_of_data -> {Enum.reverse(acc), payload}
+    end
```
Comment (mickel8) [inline lib/membrane_opus/parser.ex:70]: I would make some renamings:
`frame_size` -> `frame_len`
`frame_duration` -> `frame_size`
`frames` -> `frames_count`
`padding` -> `padding_len` or `pad_len`
`rest` -> `frames_with_payload`

also at some point `data` probably becames something like `remaining_packets` so I would  do
```
parse_packets(remaining_packets, [{new_header <> body, frame_duration * frames} | acc])
```
and by `acc` we probably mean `parsed_packets` so I would rename it too. `new_header` could be just `header` I think.

I also quite don't get why do we need 
```
{:ok, _preserved_frames_size, rest} <-
           PacketUtils.skip_frame_sizes(mode, data, max(0, frames - 1)),
``` 
and why we are passing `max(0, frames-1)`. Shouldn't we just invoke `PacketUtils.skip_frame_size(mode, data, frames)`

File: lib/membrane_opus/parser.ex [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+  def_input_pad :input,
+    demand_unit: :buffers,
+    caps: [{Opus, self_delimiting?: true}, {Stream, content: one_of([Opus, nil])}]
+
+  def_output_pad :output, caps: {Opus, self_delimiting?: false}
+
+  @impl true
+  def handle_init(_opts) do
+    {:ok, %{acc_payload: <<>>, timestamp: 0}}
```
Comment (mickel8) [inline lib/membrane_opus/parser.ex:20]: I would add some comment or docs what `acc_payload` mean

File: lib/membrane_opus/serializer.ex [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  @impl true
+  def handle_process(:input, buffer, _ctx, state) do
+    %Buffer{payload: payload} = buffer
+    {:ok, %{code: code}, data} = PacketUtils.skip_toc(payload)
+    {:ok, mode, frames, padding, data} = PacketUtils.skip_code(code, data)
+    {:ok, frames_size, body} = PacketUtils.skip_frame_sizes(mode, data, max(0, frames - 1))
+    header_size = byte_size(payload) - byte_size(body)
+    <<header::binary-size(header_size), _rest::binary>> = payload
+    last_frame_size = PacketUtils.encode_frame_size(byte_size(body) - frames_size - padding)
+    buffer = %Buffer{buffer | payload: header <> last_frame_size <> body}
```
Comment (mickel8) [inline lib/membrane_opus/serializer.ex:45]: Why do we encode here only size of the last frame insted of something like `header <> N1 <> N2 <> ... <> N[M-1] <>  N[M] <> body`?

**Review by mickel8 — CHANGES_REQUESTED** [review]

File: bundlex.exs [hunk 4]
```diff
@@ -10,14 +10,13 @@ defmodule Membrane.Opus.BundlexProject do
...
       decoder: [
-        deps: [membrane_common_c: :membrane, unifex: :unifex],
         sources: ["decoder.c"],
         libs: ["opus"],
         interface: :nif,
         preprocessor: Unifex
       ],
       encoder: [
-        deps: [membrane_common_c: :membrane, unifex: :unifex],
+        deps: [membrane_common_c: :membrane],
```
[the remark this answers is not in the document]
Comment (mat-hek) [reply bundlex.exs:19]: I assumed if it's already there, we should leave it till we reimplement native logging

File: lib/membrane_opus/packet_utils.ex [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    25 => {:celt, :super_wide, 5_000},
+    26 => {:celt, :super_wide, 10_000},
+    27 => {:celt, :super_wide, 20_000},
+    28 => {:celt, :full, 2_500},
+    29 => {:celt, :full, 5_000},
+    30 => {:celt, :full, 10_000},
+    31 => {:celt, :full, 20_000}
+  }
+
+  def skip_toc(<<config::5, stereo_flag::1, code::2, data::binary>>) do
```
[the remark this answers is not in the document]
Comment (mat-hek) [reply lib/membrane_opus/packet_utils.ex:40]: 'would' is the key here. We'll create them when implementing the 'main' parser, what will happen when we need such parser, what may never happen ;)

File: lib/membrane_opus/encoder.ex [hunk 6]
```diff
@@ -66,6 +67,12 @@ defmodule Membrane.Opus.Encoder do
     end
   end
 
+  @impl true
+  def handle_caps(:input, caps, _ctx, %{input_caps: nil} = state) do
+    output_caps = %Opus{channels: caps.channels}
```
[the remark this answers is not in the document]
Comment (mat-hek) [reply lib/membrane_opus/encoder.ex:72]: It's removed in https://github.com/membraneframework/membrane_opus_format/pull/5

File: lib/membrane_opus/packet_utils.ex [hunk 7]
```diff
@@ -0,0 +1,107 @@
+defmodule Membrane.Opus.PacketUtils do
```
[the remark this answers is not in the document]
Comment (mat-hek) [reply lib/membrane_opus/packet_utils.ex:1]: Why? It's used by both parser and serializer

File: lib/membrane_opus/parser.ex (hunk 2 above)
Comment (mat-hek) [reply lib/membrane_opus/parser.ex:20]: I think it's pretty obvious, maybe we should change the name instead?

File: lib/membrane_opus/serializer.ex (hunk 3 above)
Comment (mat-hek) [reply lib/membrane_opus/serializer.ex:45]: The previous sizes are already in the header

File: lib/membrane_opus/packet_utils.ex (hunk 7 above)
[the remark this answers is not in the document]
Comment (mickel8) [reply lib/membrane_opus/packet_utils.ex:1]: Oh, that's true, my bad :+1:

File: lib/membrane_opus/parser.ex (hunk 2 above)
Comment (mickel8) [reply lib/membrane_opus/parser.ex:20]: Yeah, I see now. I think it's ok or maybe `parsed_payload`? But I don't insist.

File: lib/membrane_opus/serializer.ex (hunk 3 above)
Comment (mickel8) [reply lib/membrane_opus/serializer.ex:45]: That's true I missed it

File: lib/membrane_opus/parser.ex (hunk 1 above)
Comment (mat-hek) [reply lib/membrane_opus/parser.ex]: I changed some names, some I left as I consider them meaningful, despite they may not conform to RFC. In particular, I think calling something 'size', when it actually amounts of time, is quite counterintuitive. Also extended moduledocs, so they shortly explain what happens here. Let me know if something is still unclear

**Review by mickel8 — COMMENTED** [review]
I have merged examples so maybe we should mention somewhere that there is possibility to use `Serializer` and `Parser`.

**Review by mickel8 — APPROVED** [review]

**Pull request merged**